src/modinv32_impl.h

```diff
@@ -168,68 +168,71 @@ typedef struct {
     int32_t u, v, q, r;
 } secp256k1_modinv32_trans2x2;
 
-/* Compute the transition matrix and zeta for 30 divsteps.
+/* Compute the transition matrix and theta for 30 divsteps (where theta = delta-1/2).
  *
- * Input:  zeta: initial zeta
- *         f0:   bottom limb of initial f
- *         g0:   bottom limb of initial g
+ * Input:  theta: initial theta
+ *         f0:    bottom limb of initial f
+ *         g0:    bottom limb of initial g
  * Output: t: transition matrix
- * Return: final zeta
+ * Return: final theta
  *
  * Implements the divsteps_n_matrix function from the explanation.
  */
-static int32_t secp256k1_modinv32_divsteps_30(int32_t zeta, uint32_t f0, uint32_t g0, secp256k1_modinv32_trans2x2 *t) {
-    /* u,v,q,r are the elements of the transformation matrix being built up,
-     * starting with the identity matrix. Semantically they are signed integers
-     * in range [-2^30,2^30], but here represented as unsigned mod 2^32. This
-     * permits left shifting (which is UB for negative numbers). The range
-     * being inside [-2^31,2^31) means that casting to signed works correctly.
-     */
-    uint32_t u = 1, v = 0, q = 0, r = 1;
-    uint32_t c1, c2, f = f0, g = g0, x, y, z;
+static int32_t secp256k1_modinv32_divsteps_30(int32_t theta, uint32_t f0, uint32_t g0, secp256k1_modinv32_trans2x2 *t) {
+    /* The transformation matrix being built up, scaled by 2^30. */
+    int32_t u = (int32_t)1 << 30, v = 0, q = 0, r = (int32_t)1 << 30, y, z;
+    uint32_t c1, c2, f = f0, g = g0, x;
     int i;
 
     for (i = 0; i < 30; ++i) {
-        VERIFY_CHECK((f & 1) == 1); /* f must always be odd */
-        VERIFY_CHECK((u * f0 + v * g0) == f << i);
-        VERIFY_CHECK((q * f0 + r * g0) == g << i);
-        /* Compute conditional masks for (zeta < 0) and for (g & 1). */
-        c1 = zeta >> 31;
+        /* f must always be odd */
+        VERIFY_CHECK((f & 1) == 1);
+        /* Minimum trailing zeros count for matrix elements decreases in each iteration */
+        VERIFY_CHECK(!((u | v | q | r) & (0xFFFFFFFFULL >> (i + 2))));
+        /* Applying the matrix so far to the initial f,g gives current f,g. */
+        VERIFY_CHECK((u >> (30 - i)) * f0 + (v >> (30 - i)) * g0 == f << i);
+        VERIFY_CHECK((q >> (30 - i)) * f0 + (r >> (30 - i)) * g0 == g << i);
```

Comment:
Just wondering, why don't these (and the analogous checks in `modinv64_impl.h`) cast to higher precision, like line 234 below?

Reply:
They are equalities modulo 2^32 (resp. 2^64) only, because that's the precision we are given f0 and g0 to. Technically maybe we only even care about modulo 2^30 (resp. 2^62), but the extra bits are there and should agree anyway. If divsteps were being done at full precision (i.e. directly on the full-size F,G) the equality would still hold, but a significant feature of safegcd is to be able to operate in limited precision most of the time, only updating the full F,G values when we need more bits to work on.

The latter matrix determinant check is full precision because, well, we're checking the exact value and not making any assumptions that we got u/v/q/r right.

Reply:
Thanks! Yes, the determinant check doesn't involve the low-order-only `f0`, `g0`, that makes sense.